File: src/templates/skills/rl3-brand/scripts/generate_pptx.py
"""
generate_pptx.py — Generate an RL3-branded PowerPoint presentation from structured content.

CLI: python generate_pptx.py --content content.json --output output.pptx
"""

import sys
import os
import argparse
import json
import zipfile
import io
import re
import tempfile

sys.path.insert(0, os.path.dirname(__file__))

from pptx import Presentation
from pptx.util import Inches, Pt, Emu
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN

import brand_tokens as bt

# ── Brand constants ────────────────────────────────────────────────

BG_COLOR = RGBColor(0x0A, 0x0A, 0x0B)
WHITE = RGBColor(0xFF, 0xFF, 0xFF)
GOLD = RGBColor(0xC8, 0xB8, 0x8A)
GRAY = RGBColor(0x7A, 0x7A, 0x7A)

FONT_HEADLINES = bt.FONTS["headlines"]   # Space Grotesk
FONT_MONO = bt.FONTS["mono"]            # Space Mono
FONT_BODY = bt.FONTS["body"]            # Instrument Sans

SLIDE_WIDTH = Inches(13.333)
SLIDE_HEIGHT = Inches(7.5)

FIXED_TIMESTAMP = "2026-01-01T00:00:00Z"


# ── Helpers ────────────────────────────────────────────────────────

def _set_slide_bg(slide):
    """Apply the standard RL3 dark background to a slide."""
    bg = slide.background
    fill = bg.fill
    fill.solid()
    fill.fore_color.rgb = BG_COLOR


def _add_textbox(slide, left, top, width, height):
    """Add a textbox and return it."""
    return slide.shapes.add_textbox(left, top, width, height)


def _add_run(paragraph, text, font_name, font_size, color, bold=False, italic=False):
    """Add a formatted run to a paragraph."""
    run = paragraph.add_run()
    run.text = text
    run.font.name = font_name
    run.font.size = font_size
    run.font.color.rgb = color
    run.font.bold = bold
    run.font.italic = italic
    return run


# ── Validation ─────────────────────────────────────────────────────

def _validate_content(content):
    """Validate required fields, raise ValueError if missing."""
    if "title" not in content:
        raise ValueError("Missing required field: 'title'")
    if "sections" not in content:
        raise ValueError("Missing required field: 'sections'")
    if not isinstance(content["sections"], list) or len(content["sections"]) < 1:
        raise ValueError("'sections' must be a non-empty list")
    for i, section in enumerate(content["sections"]):
        for field in ("number", "label", "heading", "body"):
            if field not in section:
                raise ValueError(f"Section {i}: missing required field '{field}'")


# ── Slide builders ─────────────────────────────────────────────────

def _build_title_slide(prs, content):
    """Build the title slide with RL3 logo and subtitle."""
    slide_layout = prs.slide_layouts[6]  # blank
    slide = prs.slides.add_slide(slide_layout)
    _set_slide_bg(slide)

    # Logo: "RL" white + "3" gold
    logo_box = _add_textbox(
        slide,
        Inches(4.5), Inches(2.0),
        Inches(4.333), Inches(1.5),
    )
    tf = logo_box.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    p.alignment = PP_ALIGN.CENTER
    _add_run(p, "RL", FONT_HEADLINES, Pt(72), WHITE, bold=True)
    _add_run(p, "3", FONT_HEADLINES, Pt(72), GOLD, bold=True)

    # "AI AGENCY" subtitle
    agency_box = _add_textbox(
        slide,
        Inches(4.5), Inches(3.5),
        Inches(4.333), Inches(0.6),
    )
    tf2 = agency_box.text_frame
    tf2.word_wrap = True
    p2 = tf2.paragraphs[0]
    p2.alignment = PP_ALIGN.CENTER
    _add_run(p2, "AI AGENCY", FONT_MONO, Pt(14), GRAY)

    # Content subtitle if present
    if content.get("subtitle"):
        sub_box = _add_textbox(
            slide,
            Inches(3.5), Inches(4.3),
            Inches(6.333), Inches(0.6),
        )
        tf3 = sub_box.text_frame
        tf3.word_wrap = True
        p3 = tf3.paragraphs[0]
        p3.alignment = PP_ALIGN.CENTER
        _add_run(p3, content["subtitle"], FONT_BODY, Pt(18), WHITE)

    return slide


def _build_divider_slide(prs, section):
    """Build a section divider slide with large number and heading."""
    slide_layout = prs.slide_layouts[6]
    slide = prs.slides.add_slide(slide_layout)
    _set_slide_bg(slide)

    # Large section number
    num_box = _add_textbox(
        slide,
        Inches(1.5), Inches(2.0),
        Inches(3.0), Inches(2.0),
    )
    tf = num_box.text_frame
    p = tf.paragraphs[0]
    _add_run(p, section["number"], FONT_HEADLINES, Pt(96), GOLD, bold=True)

    # Section label
    label_box = _add_textbox(
        slide,
        Inches(1.5), Inches(4.0),
        Inches(6.0), Inches(0.5),
    )
    tf2 = label_box.text_frame
    p2 = tf2.paragraphs[0]
    _add_run(p2, section["label"].upper(), FONT_MONO, Pt(12), GRAY)

    # Section heading
    heading_box = _add_textbox(
        slide,
        Inches(1.5), Inches(4.6),
        Inches(10.0), Inches(1.2),
    )
    tf3 = heading_box.text_frame
    tf3.word_wrap = True
    p3 = tf3.paragraphs[0]
    _add_run(p3, section["heading"], FONT_HEADLINES, Pt(36), WHITE, bold=True)

    return slide


def _build_content_slide(prs, section):
    """Build a section content slide with body, items, and optional callout."""
    slide_layout = prs.slide_layouts[6]
    slide = prs.slides.add_slide(slide_layout)
    _set_slide_bg(slide)

    # Body text
    body_box = _add_textbox(
        slide,
        Inches(1.5), Inches(1.5),
        Inches(10.0), Inches(1.5),
    )
    tf = body_box.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    _add_run(p, section["body"], FONT_BODY, Pt(18), WHITE)

    # Items as bullet list with arrow prefix
    items = section.get("items", [])
    if items:
        items_box = _add_textbox(
            slide,
            Inches(1.5), Inches(3.2),
            Inches(10.0), Inches(2.5),
        )
        tf2 = items_box.text_frame
        tf2.word_wrap = True
        for idx, item in enumerate(items):
            if idx == 0:
                p2 = tf2.paragraphs[0]
            else:
                p2 = tf2.add_paragraph()
            p2.space_after = Pt(6)
            _add_run(p2, f"\u2192 {item}", FONT_BODY, Pt(16), WHITE)

    # Callout if present
    if section.get("callout"):
        callout_box = _add_textbox(
            slide,
            Inches(1.5), Inches(5.5),
            Inches(10.0), Inches(0.8),
        )
        tf3 = callout_box.text_frame
        tf3.word_wrap = True
        p3 = tf3.paragraphs[0]
        _add_run(p3, section["callout"], FONT_BODY, Pt(16), GOLD, italic=True)

    return slide


def _build_final_slide(prs):
    """Build the closing slide with the RL3 cycle and tagline."""
    slide_layout = prs.slide_layouts[6]
    slide = prs.slides.add_slide(slide_layout)
    _set_slide_bg(slide)

    # Cycle label
    cycle_box = _add_textbox(
        slide,
        Inches(3.0), Inches(2.8),
        Inches(7.333), Inches(1.0),
    )
    tf = cycle_box.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    p.alignment = PP_ALIGN.CENTER
    _add_run(p, bt.CYCLE_LABEL, FONT_HEADLINES, Pt(36), WHITE, bold=True)

    # Tagline
    tag_box = _add_textbox(
        slide,
        Inches(3.0), Inches(4.0),
        Inches(7.333), Inches(0.8),
    )
    tf2 = tag_box.text_frame
    tf2.word_wrap = True
    p2 = tf2.paragraphs[0]
    p2.alignment = PP_ALIGN.CENTER
    _add_run(p2, bt.TAGLINE, FONT_BODY, Pt(16), GRAY, italic=True)

    return slide


# ── Determinism: strip timestamps from pptx ZIP ───────────────────

def _strip_timestamps(pptx_path):
    """
    Re-write the pptx ZIP so that docProps/core.xml has fixed timestamps
    and all ZIP entries have a consistent date, ensuring SHA-256 determinism.
    """
    buf = io.BytesIO()
    with open(pptx_path, "rb") as f:
        buf.write(f.read())
    buf.seek(0)

    out_buf = io.BytesIO()
    fixed_date = (2026, 1, 1, 0, 0, 0)

    with zipfile.ZipFile(buf, "r") as zin:
        with zipfile.ZipFile(out_buf, "w", zipfile.ZIP_DEFLATED) as zout:
            for item in zin.infolist():
                data = zin.read(item.filename)

                # Fix timestamps in core.xml
                if item.filename == "docProps/core.xml":
                    text = data.decode("utf-8")
                    # Replace dcterms:created and dcterms:modified values
                    text = re.sub(
                        r"(<dcterms:created[^>]*>)[^<]*(</dcterms:created>)",
                        rf"\g<1>{FIXED_TIMESTAMP}\g<2>",
                        text,
                    )
                    text = re.sub(
                        r"(<dcterms:modified[^>]*>)[^<]*(</dcterms:modified>)",
                        rf"\g<1>{FIXED_TIMESTAMP}\g<2>",
                        text,
                    )
                    data = text.encode("utf-8")

                # Set consistent date_time on every entry
                info = zipfile.ZipInfo(item.filename, date_time=fixed_date)
                info.compress_type = zipfile.ZIP_DEFLATED
                # Preserve external attributes (directory flag etc.)
                info.external_attr = item.external_attr
                zout.writestr(info, data)

    with open(pptx_path, "wb") as f:
        f.write(out_buf.getvalue())


# ── Public API ─────────────────────────────────────────────────────

def generate_pptx(content: dict, output_path: str):
    """
    Generate an RL3-branded PPTX presentation from structured content.

    Args:
        content: dict with keys title, sections (list), optional subtitle/date/footer_text
        output_path: filesystem path for the resulting .pptx file
    """
    _validate_content(content)

    prs = Presentation()
    prs.slide_width = SLIDE_WIDTH
    prs.slide_height = SLIDE_HEIGHT

    # 1. Title slide
    _build_title_slide(prs, content)

    # 2. Per section: divider + content slides
    for section in content["sections"]:
        _build_divider_slide(prs, section)
        _build_content_slide(prs, section)

    # 3. Final slide
    _build_final_slide(prs)

    # Save
    prs.save(output_path)

    # Strip timestamps for determinism
    _strip_timestamps(output_path)


# ── CLI ────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(
        description="Generate an RL3-branded PPTX from a JSON content file."
    )
    parser.add_argument("--content", required=True, help="Path to JSON content file")
    parser.add_argument("--output", required=True, help="Output .pptx file path")
    args = parser.parse_args()

    if not os.path.isfile(args.content):
        print(f"Error: content file not found: {args.content}", file=sys.stderr)
        sys.exit(1)

    with open(args.content, "r", encoding="utf-8") as f:
        content = json.load(f)

    generate_pptx(content, args.output)
    print(f"Generated: {args.output}")


if __name__ == "__main__":
    main()

File: src/templates/skills/rl3-brand/scripts/ts/generate_pptx.ts
#!/usr/bin/env npx tsx
/**
 * generate_pptx.ts — RL3 AI Agency branded PPTX generator (pptxgenjs, DEFAULT runtime).
 *
 * Usage:
 *   npx tsx generate_pptx.ts --content content.json --output output.pptx
 *
 * RL3 visual identity: dark backgrounds, gold accent (#c8b88a), Space Grotesk headlines
 */
import PptxGenJSDefault from "pptxgenjs";
// CJS/ESM interop: pptxgenjs ships as CommonJS
const PptxGenJS = ((PptxGenJSDefault as unknown as { default: typeof PptxGenJSDefault }).default ?? PptxGenJSDefault) as typeof PptxGenJSDefault;
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import * as bt from "./brand_tokens.js";

interface Section {
  number: string;
  label: string;
  heading: string;
  body: string;
  items?: string[];
  callout?: string;
}

interface Content {
  title: string;
  subtitle?: string;
  author?: string;
  sections: Section[];
}

// ---------------------------------------------------------------------------
// Slide builders — RL3 dark theme
// ---------------------------------------------------------------------------

function buildTitleSlide(pres: PptxGenJS, content: Content): void {
  const slide = pres.addSlide();
  slide.background = { color: bt.hex("background_dark") };

  // Gold accent top bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 13.333, h: 0.05,
    fill: { color: bt.hex("accent") },
    line: { color: bt.hex("accent"), width: 0 },
  });

  // RL3 wordmark
  slide.addText("RL3", {
    x: 0.5, y: 0.3, w: 2, h: 0.5,
    fontSize: 22,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
  });

  // Subtitle label in gold
  slide.addText("AI AGENCY", {
    x: 0.5, y: 0.85, w: 3, h: 0.3,
    fontSize: 9,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_body"],
    charSpacing: 5,
  });

  // Main title
  slide.addText(content.title, {
    x: 0.5, y: 2.2, w: 11, h: 2.5,
    fontSize: 44,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });

  // Subtitle
  if (content.subtitle) {
    slide.addText(content.subtitle, {
      x: 0.5, y: 4.9, w: 10, h: 0.8,
      fontSize: 18,
      color: bt.hex("text_secondary"),
      fontFace: bt.FONTS["pptx_body"],
    });
  }

  // Author footer
  const footer = [
    content.author ?? "",
    "Observar · Actuar · Iterar",
  ].filter(Boolean).join("  ·  ");
  slide.addText(footer, {
    x: 0.5, y: 6.8, w: 12, h: 0.4,
    fontSize: 11,
    color: bt.hex("text_secondary"),
    fontFace: bt.FONTS["pptx_body"],
  });
}

function buildDividerSlide(pres: PptxGenJS, section: Section): void {
  const slide = pres.addSlide();
  slide.background = { color: bt.hex("primary_light") };

  // Gold accent left bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.05, h: 7.5,
    fill: { color: bt.hex("accent") },
    line: { color: bt.hex("accent"), width: 0 },
  });

  // Large section number (decorative)
  slide.addText(section.number, {
    x: 0.5, y: 1, w: 5, h: 2.5,
    fontSize: 110,
    bold: true,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_headlines"],
    alpha: 20,
  });

  // Section heading
  slide.addText(section.heading, {
    x: 0.5, y: 3.8, w: 11, h: 1.5,
    fontSize: 32,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });

  // Section label
  slide.addText(`${section.number} — ${section.label.toUpperCase()}`, {
    x: 0.5, y: 5.5, w: 10, h: 0.4,
    fontSize: 11,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_body"],
    charSpacing: 3,
  });
}

function buildContentSlide(pres: PptxGenJS, section: Section): void {
  const slide = pres.addSlide();
  slide.background = { color: bt.hex("background_dark") };

  // Gold accent left bar
  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.05, h: 7.5,
    fill: { color: bt.hex("accent") },
    line: { color: bt.hex("accent"), width: 0 },
  });

  // Section label
  slide.addText(`${section.number} — ${section.label.toUpperCase()}`, {
    x: 0.4, y: 0.25, w: 10, h: 0.3,
    fontSize: 10,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_body"],
    charSpacing: 3,
  });

  // Heading
  slide.addText(section.heading, {
    x: 0.4, y: 0.65, w: 12, h: 1,
    fontSize: 28,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
    wrap: true,
  });

  // Divider line
  slide.addShape(pres.ShapeType.line, {
    x: 0.4, y: 1.75, w: 12.4, h: 0,
    line: { color: bt.hex("primary_mid"), width: 1 },
  });

  // Body text
  slide.addText(section.body, {
    x: 0.4, y: 1.95, w: 7.5, h: 2.5,
    fontSize: 14,
    color: bt.hex("text_primary"),
    fontFace: bt.FONTS["pptx_body"],
    lineSpacingMultiple: 1.5,
    wrap: true,
  });

  // Bullet items
  if (section.items && section.items.length > 0) {
    const bullets = section.items.map((item) => ({
      text: `› ${item}`,
      options: {
        fontSize: 13,
        color: bt.hex("text_primary"),
        fontFace: bt.FONTS["pptx_body"],
      },
    }));
    slide.addText(bullets, {
      x: 8.2, y: 1.95, w: 4.8, h: 4.5,
      wrap: true,
    });
  }

  // Callout box (gold border)
  if (section.callout) {
    slide.addShape(pres.ShapeType.rect, {
      x: 0.4, y: 4.7, w: 7.5, h: 1.5,
      fill: { color: bt.hex("accent"), alpha: 8 },
      line: { color: bt.hex("accent"), width: 1 },
    });
    slide.addText(section.callout, {
      x: 0.7, y: 4.9, w: 7, h: 1.1,
      fontSize: 13,
      italic: true,
      color: bt.hex("accent"),
      fontFace: bt.FONTS["pptx_body"],
      wrap: true,
    });
  }
}

function buildClosingSlide(pres: PptxGenJS, content: Content): void {
  const slide = pres.addSlide();
  slide.background = { color: bt.hex("background_dark") };

  slide.addShape(pres.ShapeType.rect, {
    x: 0, y: 0, w: 0.05, h: 7.5,
    fill: { color: bt.hex("accent") },
    line: { color: bt.hex("accent"), width: 0 },
  });

  slide.addText("RL3", {
    x: 0.4, y: 1.5, w: 5, h: 2,
    fontSize: 80,
    bold: true,
    color: bt.hex("white"),
    fontFace: bt.FONTS["pptx_headlines"],
  });

  slide.addText("Observar · Actuar · Iterar", {
    x: 0.4, y: 3.7, w: 10, h: 0.6,
    fontSize: 16,
    color: bt.hex("accent"),
    fontFace: bt.FONTS["pptx_body"],
  });

  slide.addText(content.title, {
    x: 0.4, y: 4.5, w: 10, h: 0.5,
    fontSize: 14,
    color: bt.hex("text_secondary"),
    fontFace: bt.FONTS["pptx_body"],
  });
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

export function generatePptx(content: Content, outputPath: string): void {
  const pres = new PptxGenJS();
  pres.layout = "LAYOUT_WIDE";
  pres.title  = content.title;
  pres.author = content.author ?? "RL3 AI Agency";

  buildTitleSlide(pres, content);
  for (const section of content.sections) {
    buildDividerSlide(pres, section);
    buildContentSlide(pres, section);
  }
  buildClosingSlide(pres, content);

  pres.writeFile({ fileName: outputPath });
  console.log(`PPTX written: ${outputPath} (${content.sections.length} sections)`);
}

function parseCli(): { content: Content; output: string } {
  const args = process.argv.slice(2);
  const getArg = (flag: string): string | undefined => {
    const idx = args.indexOf(flag);
    return idx !== -1 ? args[idx + 1] : undefined;
  };
  const contentFile = getArg("--content");
  const output = getArg("--output") ?? "output.pptx";
  if (contentFile) {
    return { content: JSON.parse(readFileSync(contentFile, "utf-8")) as Content, output };
  }
  throw new Error("Usage: npx tsx generate_pptx.ts --content content.json --output output.pptx");
}

if (process.argv[1] && fileURLToPath(import.meta.url).endsWith(process.argv[1].split("/").pop()!)) {
  const { content, output } = parseCli();
  generatePptx(content, output);
}
